packages/storage-plus/src/de.rs

```diff
@@ -0,0 +1,171 @@
+use std::array::TryFromSliceError;
+use std::convert::TryInto;
+
+use cosmwasm_std::{Addr, StdError, StdResult};
+
+use crate::keys::{IntKey, TimestampKey};
+
+pub trait KeyDeserialize {
+    type Output: Sized;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output>;
+}
+
+impl KeyDeserialize for () {
+    type Output = ();
+
+    fn from_slice(_value: &[u8]) -> StdResult<Self::Output> {
+        Ok(())
+    }
+}
+
+impl KeyDeserialize for Vec<u8> {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        Ok(value.to_vec())
+    }
+}
+
+impl KeyDeserialize for &Vec<u8> {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for [u8] {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for &[u8] {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for String {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_utf8(value.to_vec())
+            // FIXME: Add and use StdError utf-8 error From helper
+            .map_err(|err| StdError::generic_err(err.to_string()))
+    }
+}
+
+impl KeyDeserialize for &String {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for str {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for &str {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for Addr {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for &Addr {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+macro_rules! integer_de {
+    (for $($t:ty),+) => {
+        $(impl KeyDeserialize for IntKey<$t> {
+            type Output = $t;
+
+            fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+                Ok(<$t>::from_be_bytes(value.try_into()
```

Comment:
This will fail if it is too short, right?
But not if it is too long?

Since you have a space to return errors in parsing (good), how about an explicit length check before this line?

Reply:
Interestingly, `try_into` will err if the slice is too long too. Which is something I just checked, and didn't knew when writing this.